src/nbl/video/IDescriptorPool.cpp

```diff
@@ -4,35 +4,67 @@
-        if (count == 0ull)
-            continue;
+        auto* ds = static_cast<IGPUDescriptorSet*>(pDescriptorWrites[i].dstSet);
+        ds->incrementVersion();
 
-        if (m_flags & ECF_FREE_DESCRIPTOR_SET_BIT)
-            offsets.data[type] = m_generalAllocators[type].alloc_addr(count, 1u);
-        else
-            offsets.data[type] = m_linearAllocators[type].alloc_addr(count, 1u);
+        auto* descriptors = ds->getDescriptors(pDescriptorWrites[i].descriptorType, pDescriptorWrites[i].binding);
+        auto* samplers = ds->getMutableSamplers(pDescriptorWrites[i].binding);
+        for (auto j = 0; j < pDescriptorWrites[i].count; ++j)
+        {
+            descriptors[j] = pDescriptorWrites[i].info[j].desc;
 
-        assert((offsets.data[type] < m_maxDescriptorCount[i]) && "PANIC: Allocation failed. This shoudn't have happened!");
+            if (samplers)
+                samplers[j] = pDescriptorWrites[i].info[j].info.image.sampler;
+        }
     }
 
-    const auto mutableSamplerCount = layout->getTotalMutableSamplerCount();
-    if (mutableSamplerCount != 0ull)
+    for (auto i = 0; i < descriptorCopyCount; ++i)
     {
-        if (m_flags & ECF_FREE_DESCRIPTOR_SET_BIT)
-            offsets.data[asset::EDT_COUNT] = m_generalAllocators[asset::EDT_COUNT].alloc_addr(mutableSamplerCount, 1u);
-        else
-            offsets.data[asset::EDT_COUNT] = m_linearAllocators[asset::EDT_COUNT].alloc_addr(mutableSamplerCount, 1u);
+        const auto* srcDS = static_cast<const IGPUDescriptorSet*>(pDescriptorCopies[i].srcSet);
+        auto* dstDS = static_cast<IGPUDescriptorSet*>(pDescriptorCopies[i].dstSet);
+
+        auto foundBindingInfo = std::lower_bound(srcDS->getLayout()->getBindings().begin(), srcDS->getLayout()->getBindings().end(), pDescriptorCopies[i].srcBinding,
+            [](const IGPUDescriptorSetLayout::SBinding& a, const uint32_t b) -> bool
+            {
+                return a.binding < b;
+            });
+
+        if (foundBindingInfo->binding != pDescriptorCopies[i].srcBinding)
+            return false;
+
+        const asset::E_DESCRIPTOR_TYPE descriptorType = foundBindingInfo->type;
+
+        auto* srcDescriptors = srcDS->getDescriptors(descriptorType, pDescriptorCopies[i].srcBinding);
+        auto* srcSamplers = srcDS->getMutableSamplers(pDescriptorCopies[i].srcBinding);
+        if (!srcDescriptors)
+            return false;
+
+        auto* dstDescriptors = dstDS->getDescriptors(descriptorType, pDescriptorCopies[i].dstBinding);
+        auto* dstSamplers = dstDS->getMutableSamplers(pDescriptorCopies[i].dstBinding);
+        if (!dstDescriptors)
+            return false;
+
+        memcpy(dstDescriptors, srcDescriptors, pDescriptorCopies[i].count * sizeof(core::smart_refctd_ptr<const asset::IDescriptor>));
+
+        if (srcSamplers && dstSamplers)
+            memcpy(dstSamplers, srcSamplers, pDescriptorCopies[i].count * sizeof(core::smart_refctd_ptr<const IGPUSampler>));
```

Comment:
no it will do nothing of the kind, `memcpy` does nothing to call ctors or dtors.

Our copy ctor does, **so you must use `std::copy_n`**

You will leak the old descriptor (because no dtor called on smart pointer) and crash due to new descriptor being destroyed due to a missing constructor call.